### DISCUSSION TITLE
Loading a YAML Representation Graph

### DISCUSSION THREAD

**Discussion by colltoaction** [networkx/networkx#6808] [Ideas]
Hi, I'm starting this package and wanted to get your early feedback.

The idea is to load the YAML representation graph into NetworkX for manipulation. Then dump the NetworkX objects into a YAML document with a friendly syntax.

https://github.com/colltoaction/nx_yaml

Here are two examples that are working so far: 

```yaml
source: source
```

![loop](https://github.com/networkx/networkx/assets/1548532/d45d495c-a000-4712-8ce4-9b031f5f00d2)

```yaml
tree root:
    left:
    right:
```

![two_leaf_tree](https://github.com/networkx/networkx/assets/1548532/04773dc8-0deb-4446-bd4f-552b047dfb67)

**Comment by MridulS** [comment]
Thanks for opening the discussion!

Btw we used to have a yaml conversion https://github.com/networkx/networkx/blob/networkx-2.5/networkx/readwrite/nx_yaml.py but we deprecated that as it was just a wrapper around:

``` python
import yaml
with open('path_to_yaml_file', 'w') as fh:
    yaml.dump(G, fh)
with open('path_to_yaml_file', 'r') as fh:
    G = yaml.load(fh, Loader=yaml.Loader)
```

Is there any specific data representation for graphs in YAML that you are aiming for?

**Reply by colltoaction** [reply]
Yes, I've seen that! This would be very different from that approach because it is based on the native data structures instead of the representation graph.

I studied the yaml spec and think it is possible to extract a document's underlying representation graph:
* https://yaml.org/spec/1.2.2/#chapter-3-processes-and-models
* https://yaml.org/spec/1.2.2/#321-representation-graph)

<img width="1119" alt="image" src="https://github.com/networkx/networkx/assets/1548532/ba46ae01-8815-4bb0-a485-61eaf8ee5ffe">

**Reply by colltoaction** [reply]
Hi again @MridulS! I'm having an interesting conversation where I discuss if the YAML really defines a graph.

Any feedback is appreciated!

https://github.com/yaml/yaml-spec/issues/312

**Reply by rossbar** [reply]
I'll quote @Thom1729 here, as this very nicely summarizes my current thoughts as well:

> I may have lost track of exactly what you're looking to accomplish. I think that there are good solutions for encoding mathematical graphs as YAML representation graphs, and also good solutions for encoding YAML representation graphs as mathematical graphs, but due to the different data models I don't think you're going to find a reasonable bijection, or an encoding either way that never requires adds nodes/vertices — and an encoding that minimizes extra nodes/vertices may be far from the most ergonomic.

From that discussion, I gather that I missing a lot of context related to the yaml spec. Therefore I don't feel I can really comment on whether or not it makes sense to try to model components of the yaml spec with NetworkX.

If the goal is to come up with a representation for graph data *using* yaml (which is how I and likely @MridulS interpreted the OP), then I think the conversation needs to focus a bit more on motivation (and bear in mind the [wisdom of xkcd](https://xkcd.com/927/)).

**Reply by colltoaction** [reply]
I love that strip! I have considered that a lot, since often I see new programming languages with new syntaxes. I want to fight that by supercharging YAML with programming tools.

YAML spec has, as the owner explains, a graph representation. Leveraging that feature one could view, edit, ..., any .yaml in terms of graphs, separate from both the text format and native structure.
